sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1287,3 +1288,33 @@ object RemoveRepetitionFromGroupExpressions extends Rule[LogicalPlan] {
       a.copy(groupingExpressions = newGrouping)
   }
 }
+
+/**
+ * Splits [[Aggregate]] on [[Expand]], which has large number of projections,
+ * into various [[Aggregate]]s.
+ */
+object SplitAggregateWithExpand extends Rule[LogicalPlan] {
+  /**
+   * Split [[Expand]] operator to a number of [[Expand]] operators
+   */
+  private def splitExpand(expand: Expand): Seq[Expand] = {
+    val len = expand.projections.length
+    val allProjections = expand.projections
+    Seq.tabulate(len)(
+      i => Expand(Seq(allProjections(i)), expand.output, expand.child)
```

Comment:
Why not remove the Expand altogether. The only thing you you might need to do is to make sure that the expressions produce the correct attributes.

Reply:
Shall we keep Expand here for further optimization?  I think we may put more than one projections together for Union operator.

In our production, we found some cube queries may even have 22 dimensions, whch result to 2^22=4194304 projections. In such case, it is not appropriate to tansform it into "one-projection-one-child" for Union node.

Reply:
How do you propose optimizing further?

Reply:
Can we set another config param here?

Reply:
And I think we may figure out some cost-based methods.